usr/src/man/man1/ipcs.1
'\" te
.\"  Copyright 1989 AT&T
.\" Copyright (c) 2008, Sun Microsystems, Inc.  All Rights Reserved
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH IPCS 1 "Jan 17, 2008"
.SH NAME
ipcs \- report inter-process communication facilities status
.SH SYNOPSIS
.LP
.nf
\fBipcs\fR [\fB-aAbciJmopqstZ\fR] [\fB-D\fR \fImtype\fR] [\fB-z\fR \fIzone\fR]
.fi

.SH DESCRIPTION
.sp
.LP
The \fBipcs\fR utility prints information about active inter-process
communication facilities. The information that is displayed is controlled by
the options supplied. Without options, information is printed in short format
for message queues, shared memory, and semaphores that are currently active in
the system.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-m\fR\fR
.ad
.RS 6n
Prints information about active shared memory segments.
.RE

.sp
.ne 2
.na
\fB\fB-q\fR\fR
.ad
.RS 6n
Prints information about active message queues.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR\fR
.ad
.RS 6n
Prints information about active semaphores.
.RE

.sp
.LP
If \fB-m\fR, \fB-q\fR, or \fB-s\fR are specified, information about only those
indicated is printed. If none of these three is specified, information about
all three is printed subject to these options:
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.RS 12n
Uses all XCU5 print options. (This is a shorthand notation for \fB-b\fR,
\fB-c\fR, \fB-o\fR, \fB-p\fR, and \fB-t\fR.)
.RE

.sp
.ne 2
.na
\fB\fB-A\fR\fR
.ad
.RS 12n
Uses all print options. (This is a shorthand notation for \fB-b\fR, \fB-c\fR,
\fB-i\fR, \fB-J\fR, \fB-o\fR, \fB-p\fR, and \fB-t\fR.)
.RE

.sp
.ne 2
.na
\fB\fB-b\fR\fR
.ad
.RS 12n
Prints information on biggest allowable size: maximum number of bytes in
messages on queue for message queues, size of segments for shared memory, and
number of semaphores in each set for semaphores. See below for meaning of
columns in a listing.
.RE

.sp
.ne 2
.na
\fB\fB-c\fR\fR
.ad
.RS 12n
Prints creator's login name and group name. See below.
.RE

.sp
.ne 2
.na
\fB\fB-D\fR \fImtype\fR\fR
.ad
.RS 12n
Displays, in hexadecimal and \fBASCII\fR, the contents of all messages of type
\fImtype\fR found on any message queue that the user invoking \fBipcs\fR has
permission to read. If \fImtype\fR is \fB0\fR, all messages are displayed. If
\fImtype\fR is negative, all messages with type less than or equal to the
absolute value of \fImtype\fR are displayed. (See \fBmsgrcv\fR(2) and
\fBmsgsnap\fR(2)).
.RE

.sp
.ne 2
.na
\fB\fB-i\fR\fR
.ad
.RS 12n
Prints number of \fBISM\fR attaches to shared memory segments.
.RE

.sp
.ne 2
.na
\fB\fB-J\fR\fR
.ad
.RS 12n
Prints the creator's project.
.RE

.sp
.ne 2
.na
\fB\fB-o\fR\fR
.ad
.RS 12n
Prints information on outstanding usage: number of messages on queue and total
number of bytes in messages on queue for message queues and number of processes
attached to shared memory segments.
.RE

.sp
.ne 2
.na
\fB\fB-p\fR\fR
.ad
.RS 12n
Prints process number information: process \fBID\fR of last process to send a
message, process \fBID\fR of last process to receive a message on message
queues, process \fBID\fR of creating process, and process \fBID\fR of last
process to attach or detach on shared memory segments. See below.
.RE

.sp
.ne 2
.na
\fB\fB-t\fR\fR
.ad
.RS 12n
Prints time information: time of the last control operation that changed the
access permissions for all facilities, time of last \fBmsgsnd\fR(2) and last
\fBmsgrcv\fR(2) on message queues, time of last \fBshmat\fR(2) and last
\fBshmdt\fR(2) on shared memory (see \fBshmop\fR(2)), time of last
\fBsemop\fR(2) on semaphores. See below.
.RE

.sp
.ne 2
.na
\fB\fB-z\fR \fIzone\fR\fR
.ad
.RS 12n
Prints information about facilities associated with the specified zone (see
\fBzones\fR(7)). The zone can be specified as either a name or a numeric id.
The default is to display information about the zone in which the command is
executing. Notice that this option is only useful when executing in the global
zone.
.RE

.sp
.ne 2
.na
\fB\fB-Z\fR\fR
.ad
.RS 12n
When executing in the global zone, prints information about all zones.
Otherwise, prints information about the zone in which the command is executing.
The output includes the zone associated with each facility.
.RE

.sp
.LP
The column headings and the meaning of the columns in an \fBipcs\fR listing are
given below. The letters in parentheses indicate the options that cause the
corresponding heading to appear and "all" means that the heading always
appears. \fBNote:\fR These options only determine what information is provided
for each facility; they do not determine which facilities are listed.
.sp
.ne 2
.na
\fB\fBT\fR (all)\fR
.ad
.RS 19n
Type of the facility:
.sp
.ne 2
.na
\fB\fBq\fR\fR
.ad
.RS 5n
message queue
.RE

.sp
.ne 2
.na
\fB\fBm\fR\fR
.ad
.RS 5n
shared memory segment
.RE

.sp
.ne 2
.na
\fB\fBs\fR\fR
.ad
.RS 5n
semaphore
.RE

.RE

.sp
.ne 2
.na
\fB\fBID\fR (all)\fR
.ad
.RS 19n
The identifier for the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBKEY\fR (all)\fR
.ad
.RS 19n
The key used as an argument to \fBmsgget\fR(2), \fBsemget\fR(2), or
\fBshmget\fR(2) to create the facility entry. (\fBNote:\fR The  key of a shared
memory segment is changed to \fBIPC_PRIVATE\fR when the segment has been
removed until all processes attached to the segment detach it.)
.RE

.sp
.ne 2
.na
\fB\fBMODE\fR (all)\fR
.ad
.RS 19n
The facility access modes and flags: The mode consists of 11 characters that
are interpreted as follows. The first two characters are:
.sp
.ne 2
.na
\fB\fBR\fR\fR
.ad
.RS 5n
A process is waiting on a \fBmsgrcv\fR(2).
.RE

.sp
.ne 2
.na
\fB\fBS\fR\fR
.ad
.RS 5n
A process is waiting on a \fBmsgsnd\fR(2).
.RE

.sp
.ne 2
.na
\fB\fB-\fR\fR
.ad
.RS 5n
The corresponding special flag is not set.
.RE

The next nine characters are interpreted as three sets of three bits each. The
first set refers to the owner's permissions; the next to permissions of others
in the user-group of the facility entry; and the last to all others. Within
each set, the first character indicates permission to read, the second
character indicates permission to write or alter the facility entry, and the
last character is currently unused.
.sp
The permissions are indicated as follows:
.sp
.ne 2
.na
\fB\fBr\fR\fR
.ad
.RS 8n
Read permission is granted.
.RE

.sp
.ne 2
.na
\fB\fBw\fR\fR
.ad
.RS 8n
Write permission is granted.
.RE

.sp
.ne 2
.na
\fB\fBa\fR\fR
.ad
.RS 8n
Alter permission is granted.
.RE

.sp
.ne 2
.na
\fB\fB\(mi\fR\fR
.ad
.RS 8n
The indicated permission is not granted.
.RE

.RE

.sp
.ne 2
.na
\fB\fBOWNER\fR (all)\fR
.ad
.RS 19n
The login name of the owner of the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBGROUP\fR (all)\fR
.ad
.RS 19n
The group name of the group of the owner of the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBCREATOR\fR (a,A,c)\fR
.ad
.RS 19n
The login name of the creator of the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBCGROUP\fR (a,A,c)\fR
.ad
.RS 19n
The group name of the group of the creator of the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBCBYTES\fR (a,A,o)\fR
.ad
.RS 19n
The number of bytes in messages currently outstanding on the associated message
queue.
.RE

.sp
.ne 2
.na
\fB\fBQNUM\fR (a,A,o)\fR
.ad
.RS 19n
The number of messages currently outstanding on the associated message queue.
.RE

.sp
.ne 2
.na
\fB\fBQBYTES\fR (a,A,b)\fR
.ad
.RS 19n
The maximum number of bytes allowed in messages outstanding on the associated
message queue.
.RE

.sp
.ne 2
.na
\fB\fBLSPID\fR (a,A,p)\fR
.ad
.RS 19n
The process \fBID\fR of the last process to send a message to the associated
queue.
.RE

.sp
.ne 2
.na
\fB\fBLRPID\fR (a,A,p)\fR
.ad
.RS 19n
The process \fBID\fR of the last process to receive a message from the
associated queue.
.RE

.sp
.ne 2
.na
\fB\fBSTIME\fR (a,A,t)\fR
.ad
.RS 19n
The time the last message was sent to the associated queue.
.RE

.sp
.ne 2
.na
\fB\fBRTIME\fR (a,A,t)\fR
.ad
.RS 19n
The time the last message was received from the associated queue.
.RE

.sp
.ne 2
.na
\fB\fBCTIME\fR (a,A,t)\fR
.ad
.RS 19n
The time when the associated entry was created or changed.
.RE

.sp
.ne 2
.na
\fB\fBISMATTCH\fR (a,i)\fR
.ad
.RS 19n
The number of \fBISM\fR attaches to the associated shared memory segments.
.RE

.sp
.ne 2
.na
\fB\fBNATTCH\fR (a,A,o)\fR
.ad
.RS 19n
The number of processes attached to the associated shared memory segment.
.RE

.sp
.ne 2
.na
\fB\fBSEGSZ\fR (a,A,b)\fR
.ad
.RS 19n
The size of the associated shared memory segment.
.RE

.sp
.ne 2
.na
\fB\fBCPID\fR (a,A,p)\fR
.ad
.RS 19n
The process \fBID\fR of the creator of the shared memory entry.
.RE

.sp
.ne 2
.na
\fB\fBLPID\fR (a,A,p)\fR
.ad
.RS 19n
The process \fBID\fR of the last process to attach or detach the shared memory
segment.
.RE

.sp
.ne 2
.na
\fB\fBATIME\fR (a,A,t)\fR
.ad
.RS 19n
The time the last attach was completed to the associated shared memory segment.
.RE

.sp
.ne 2
.na
\fB\fBDTIME\fR (a,A,t)\fR
.ad
.RS 19n
The time the last detach was completed on the associated shared memory segment.
.RE

.sp
.ne 2
.na
\fB\fBNSEMS\fR (a,A,b)\fR
.ad
.RS 19n
The number of semaphores in the set associated with the semaphore entry.
.RE

.sp
.ne 2
.na
\fB\fBOTIME\fR (a,A,t)\fR
.ad
.RS 19n
The time the last semaphore operation was completed on the set associated with
the semaphore entry.
.RE

.sp
.ne 2
.na
\fB\fBPROJECT\fR (J,A)\fR
.ad
.RS 19n
The project name of the creator of the facility entry.
.RE

.sp
.ne 2
.na
\fB\fBZONE\fR (Z)\fR
.ad
.RS 19n
The zone with which the facility is associated.
.RE

.SH ENVIRONMENT VARIABLES
.sp
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of  \fBipcs\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.sp
.ne 2
.na
\fB\fBTZ\fR\fR
.ad
.RS 6n
Determine the timezone for the time strings written by \fBipcs\fR.
.RE

.SH FILES
.sp
.ne 2
.na
\fB\fB/etc/group\fR\fR
.ad
.RS 15n
group names
.RE

.sp
.ne 2
.na
\fB\fB/etc/passwd\fR\fR
.ad
.RS 15n
user names
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
Interface Stability	Committed
_
Standard	See \fBstandards\fR(7).
.TE

.SH SEE ALSO
.sp
.LP
.BR ipcrm (1),
.BR msgget (2),
.BR msgids (2),
.BR msgrcv (2),
.BR msgsnap (2),
.BR msgsnd (2),
.BR semget (2),
.BR semids (2),
.BR semop (2),
.BR shmctl (2),
.BR shmget (2),
.BR shmids (2),
.BR shmop (2),
.BR attributes (7),
.BR environ (7),
.BR standards (7),
.BR zones (7)
.SH NOTES
.sp
.LP
Things can change while \fBipcs\fR is running. The information it gives is
guaranteed to be accurate only when it was retrieved.


usr/src/cmd/ipcs/ipcs.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


#pragma ident	"%Z%%M%	%I%	%E% SMI"

/*
 * ipcs - IPC status
 *
 * Examine and print certain things about
 * message queues, semaphores and shared memory.
 *
 * IPC information is obtained via msgctl64, semctl64 and shmctl64.
 * As of SunOS 5.8, the IPC identifiers are obtained from msgids(),
 * semids(), and shmids() rather than reading them from /dev/kmem.
 * This ensures that the information in each msgid_ds, semid_ds or
 * shmid_ds data structure that we obtain is complete and consistent,
 * and allows us not to be a setgid-sys isaexec process.
 */

#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/ipc_impl.h>
#include <sys/msg.h>
#include <sys/sem.h>
#include <sys/shm.h>
#include <errno.h>
#include <fcntl.h>
#include <time.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <locale.h>
#include <langinfo.h>
#include <string.h>
#include <limits.h>
#include <project.h>
#include <zone.h>

#define	USAGE	\
	"usage: ipcs [-AabciJmopqstZ] [-D mtype] [-z zone]\n"

static char chdr[] = "T         ID      KEY        MODE        OWNER    GROUP";
						/* common header format */
static char chdr2[] = "  CREATOR   CGROUP";	/* c option header format */
static char chdr3[] = "         PROJECT";	/* J option header format */
static char opts[] = "AabciJmopqstD:z:Z";	/* getopt options */

static long	mtype;		/* -D: user-supplied message type */
static zoneid_t	zoneid;		/* -z: user-supplied zone id */

static int	bflg,		/* biggest size: */
				/*	segsz on m; qbytes on q; nsems on s */
		cflg,		/* creator's login and group names */
		Dflg,		/* dump contents of message queues */
		iflg,		/* ISM attaches */
		Jflg,		/* dump project name */
		mflg,		/* shared memory status */
		oflg,		/* outstanding data: */
				/*	nattch on m; cbytes, qnum on q */
		pflg,		/* process id's: lrpid, lspid on q; */
				/*	cpid, lpid on m */
		qflg,		/* message queue status */
		sflg,		/* semaphore status */
		tflg,		/* times: atime, ctime, dtime on m;	*/
				/*	ctime, rtime, stime on q;	*/
				/*	ctime, otime on s */
		zflg,		/* show only objects from specified zone */
		Zflg,		/* display zone name */
		err;		/* option error count */

static void hp(char, char *, struct ipc_perm64 *, int);
static void jp(struct ipc_perm64 *);
static void tp(ipc_time_t);
static void dumpmsgq(int);
static void dumpmsg(long, char *, size_t);
static zoneid_t getzone(char *);
static void printzone(zoneid_t);

int
main(int argc, char *argv[])
{
	static	int	*ids;	/* array of IPC identifiers from *ids() */
	static	uint_t	nids;	/* number of entries in ids */

	int	o;	/* option flag */
	int	id;	/* IPC identifier */
	int	i;
	uint_t	n;	/* table size */
	time_t	now;	/* date */
	char	tbuf[BUFSIZ];
	char	*dfmt;  /* date format pointer */
	char	*endptr;	/* terminator for strtol() */

	(void) setlocale(LC_ALL, "");
	(void) textdomain(TEXT_DOMAIN);

	(void) memset(tbuf, 0, sizeof (tbuf));
	dfmt = nl_langinfo(_DATE_FMT);

	zoneid = getzoneid();	/* default zone id if -z and -Z not used */

	/* Go through the options and set flags. */
	while ((o = getopt(argc, argv, opts)) != EOF) {
		switch (o) {
		case 'A':
			bflg = cflg = iflg = oflg = pflg = tflg = Jflg = 1;
			break;
		case 'a':
			bflg = cflg = oflg = pflg = tflg = 1;
			break;
		case 'b':
			bflg = 1;
			break;
		case 'c':
			cflg = 1;
			break;
		case 'D':
			mtype = strtol(optarg, &endptr, 0);
			if (endptr == optarg || *endptr != '\0') {
				(void) fprintf(stderr,
				    gettext("ipcs: invalid message type: %s\n"),
				    optarg);
				err++;
				break;
			}
			Dflg = 1;
			break;
		case 'i':
			iflg = 1;
			break;
		case 'J':
			Jflg = 1;
			break;
		case 'm':
			mflg = 1;
			break;
		case 'o':
			oflg = 1;
			break;
		case 'p':
			pflg = 1;
			break;
		case 'q':
			qflg = 1;
			break;
		case 's':
			sflg = 1;
			break;
		case 't':
			tflg = 1;
			break;
		case 'z':
			zflg = 1;
			zoneid = getzone(optarg);
			break;
		case 'Z':
			Zflg = 1;
			break;
		case '?':
			err++;
			break;
		}
	}
	if (err || (optind < argc)) {
		(void) fprintf(stderr, gettext(USAGE));
		exit(1);
	}

	if ((mflg + qflg + sflg) == 0)
		mflg = qflg = sflg = 1;

	now = time(NULL);
	(void) strftime(tbuf, sizeof (tbuf), dfmt, localtime(&now));
	(void) printf(gettext("IPC status from <running system> as of %s\n"),
	    tbuf);

	/*
	 * Print Message Queue status report.
	 */
	if (qflg) {
		struct msqid_ds64 qds;

		for (;;) {
			if (msgids(ids, nids, &n) != 0) {
				perror("msgids");
				exit(1);
			}
			if (n <= nids)
				break;
			ids = realloc(ids, (nids = n) * sizeof (int));
		}

		(void) printf("%s%s%s%s%s%s%s%s\n", chdr,
		    cflg ? chdr2 : "",
		    oflg ? " CBYTES  QNUM" : "",
		    bflg ? " QBYTES" : "",
		    pflg ? " LSPID LRPID" : "",
		    tflg ? "   STIME    RTIME    CTIME " : "",
		    Jflg ? chdr3 : "",
		    Zflg ? "     ZONE" : "");

		(void) printf(gettext("Message Queues:\n"));

		for (i = 0; i < n; i++) {
			id = ids[i];
			if (msgctl64(id, IPC_STAT64, &qds) < 0)
				continue;
			/* ignore zone if -Z was used and -z wasn't */
			if ((zflg || !Zflg) &&
			    qds.msgx_perm.ipcx_zoneid != zoneid)
				continue;
			hp('q', "SRrw-rw-rw-", &qds.msgx_perm, id);
			if (oflg)
				(void) printf(" %6llu %5llu",
				    qds.msgx_cbytes, qds.msgx_qnum);
			if (bflg)
				(void) printf(" %6llu", qds.msgx_qbytes);
			if (pflg)
				(void) printf(" %5d %5d",
				    (int)qds.msgx_lspid, (int)qds.msgx_lrpid);
			if (tflg) {
				tp(qds.msgx_stime);
				tp(qds.msgx_rtime);
				tp(qds.msgx_ctime);
			}
			if (Jflg)
				jp(&qds.msgx_perm);
			if (Zflg)
				printzone(qds.msgx_perm.ipcx_zoneid);
			(void) printf("\n");
			if (Dflg)
				dumpmsgq(id);
		}
	}

	/*
	 * Print Shared Memory status report.
	 */
	if (mflg) {
		struct shmid_ds64 mds;

		for (;;) {
			if (shmids(ids, nids, &n) != 0) {
				perror("shmids");
				exit(1);
			}
			if (n <= nids)
				break;
			ids = realloc(ids, (nids = n) * sizeof (int));
		}

		if (!qflg || oflg || bflg || pflg || tflg || iflg)
			(void) printf("%s%s%s%s%s%s%s%s%s\n", chdr,
			    cflg ? chdr2 : "",
			    oflg ? " NATTCH" : "",
			    bflg ? "      SEGSZ" : "",
			    pflg ? "  CPID  LPID" : "",
			    tflg ? "   ATIME    DTIME    CTIME " : "",
			    iflg ? " ISMATTCH" : "",
			    Jflg ? chdr3 : "",
			    Zflg ? "     ZONE" : "");

		(void) printf(gettext("Shared Memory:\n"));

		for (i = 0; i < n; i++) {
			id = ids[i];
			if (shmctl64(id, IPC_STAT64, &mds) < 0)
				continue;
			/* ignore zone if -Z was used and -z wasn't */
			if ((zflg || !Zflg) &&
			    mds.shmx_perm.ipcx_zoneid != zoneid)
				continue;
			hp('m', "--rw-rw-rw-", &mds.shmx_perm, id);
			if (oflg)
				(void) printf(" %6llu", mds.shmx_nattch);
			if (bflg)
				(void) printf(" %10llu", mds.shmx_segsz);
			if (pflg)
				(void) printf(" %5d %5d",
				    (int)mds.shmx_cpid, (int)mds.shmx_lpid);
			if (tflg) {
				tp(mds.shmx_atime);
				tp(mds.shmx_dtime);
				tp(mds.shmx_ctime);
			}
			if (iflg)
				(void) printf(" %8llu", mds.shmx_cnattch);
			if (Jflg)
				jp(&mds.shmx_perm);
			if (Zflg)
				printzone(mds.shmx_perm.ipcx_zoneid);
			(void) printf("\n");
		}
	}

	/*
	 * Print Semaphore facility status.
	 */
	if (sflg) {
		struct semid_ds64 sds;
		union semun {
			int val;
			struct semid_ds64 *buf;
			ushort_t *array;
		} semarg;
		semarg.buf = &sds;

		for (;;) {
			if (semids(ids, nids, &n) != 0) {
				perror("semids");
				exit(1);
			}
			if (n <= nids)
				break;
			ids = realloc(ids, (nids = n) * sizeof (int));
		}

		if (bflg || tflg || (!qflg && !mflg))
			(void) printf("%s%s%s%s%s%s\n", chdr,
			    cflg ? chdr2 : "",
			    bflg ? " NSEMS" : "",
			    tflg ? "   OTIME    CTIME " : "",
			    Jflg ? chdr3 : "",
			    Zflg ? "     ZONE" : "");

		(void) printf(gettext("Semaphores:\n"));

		for (i = 0; i < n; i++) {
			id = ids[i];
			if (semctl64(id, 0, IPC_STAT64, semarg) < 0)
				continue;
			/* ignore zone if -Z was used and -z wasn't */
			if ((zflg || !Zflg) &&
			    sds.semx_perm.ipcx_zoneid != zoneid)
				continue;
			hp('s', "--ra-ra-ra-", &sds.semx_perm, id);
			if (bflg)
				(void) printf(" %5u", sds.semx_nsems);
			if (tflg) {
				tp(sds.semx_otime);
				tp(sds.semx_ctime);
			}
			if (Jflg)
				jp(&sds.semx_perm);
			if (Zflg)
				printzone(sds.semx_perm.ipcx_zoneid);
			(void) printf("\n");
		}
	}

	return (0);
}

/*
 * hp - common header print
 */
static void
hp(char type, char *modesp, struct ipc_perm64 *permp, int slot)
{
	int		i;	/* loop control */
	struct group	*g;	/* ptr to group group entry */
	struct passwd	*u;	/* ptr to user passwd entry */
	char		keyfield[16];

	(void) snprintf(keyfield, sizeof (keyfield), "  0x%x", permp->ipcx_key);
	(void) printf("%c %10d %-13s", type, slot, keyfield);

	for (i = 02000; i; modesp++, i >>= 1)
		(void) printf("%c", (permp->ipcx_mode & i) ? *modesp : '-');
	if ((u = getpwuid(permp->ipcx_uid)) == NULL)
		(void) printf("%9d", (int)permp->ipcx_uid);
	else
		(void) printf("%9.8s", u->pw_name);
	if ((g = getgrgid(permp->ipcx_gid)) == NULL)
		(void) printf("%9d", (int)permp->ipcx_gid);
	else
		(void) printf("%9.8s", g->gr_name);

	if (cflg) {
		if ((u = getpwuid(permp->ipcx_cuid)) == NULL)
			(void) printf("%9d", (int)permp->ipcx_cuid);
		else
			(void) printf("%9.8s", u->pw_name);
		if ((g = getgrgid(permp->ipcx_cgid)) == NULL)
			(void) printf("%9d", (int)permp->ipcx_cgid);
		else
			(void) printf("%9.8s", g->gr_name);
	}
}

/*
 * jp - project header print
 */
static void
jp(struct ipc_perm64 *permp)
{
	struct project	proj;
	char		buf[PROJECT_BUFSZ];

	if ((getprojbyid(permp->ipcx_projid, &proj, buf,
	    PROJECT_BUFSZ)) == NULL)
		(void) printf("%16ld", permp->ipcx_projid);
	else
		(void) printf("%16.15s", proj.pj_name);
}

/*
 * tp - time entry printer
 */
void
tp(ipc_time_t gmt64)
{
	struct tm *t;	/* ptr to converted time */
	time_t gmt = (time_t)gmt64;

	if (gmt && gmt64 <= UINT_MAX) {
		t = localtime(&gmt);
		(void) printf(" %2d:%2.2d:%2.2d",
		    t->tm_hour, t->tm_min, t->tm_sec);
	} else {
		(void) printf("%9s", gettext(" no-entry"));
	}
}

/* Round up to a sizeof (size_t) boundary */
#define	SZROUND(x)	(((x) + sizeof (size_t) - 1) & ~(sizeof (size_t) - 1))

/*
 * dumpmsgq - dump all messages on a message queue
 */
void
dumpmsgq(int msqid)
{
	static struct msgsnap_head *buf = NULL;
	static size_t bufsize;

	struct msgsnap_mhead *mhead;
	size_t i;

	/* allocate the minimum required buffer size on first time through */
	if (buf == NULL)
		buf = malloc(bufsize = sizeof (struct msgsnap_head));

	/*
	 * Fetch all messages specified by mtype from
	 * the queue while leaving the queue intact.
	 */
	for (;;) {
		if (msgsnap(msqid, buf, bufsize, mtype) != 0) {
			/*
			 * Don't complain; either the user does not have
			 * read permission on msqid or msqid was deleted.
			 */
			return;
		}
		if (bufsize >= buf->msgsnap_size) {
			/* we collected all of the messages */
			break;
		}
		/* The buffer is too small; allocate a bigger buffer */
		buf = realloc(buf, bufsize = buf->msgsnap_size);
	}

	/*
	 * Process each message in the queue (there may be none).
	 * The first message header starts just after the buffer header.
	 */
	mhead = (struct msgsnap_mhead *)(buf + 1);
	for (i = 0; i < buf->msgsnap_nmsg; i++) {
		size_t mlen = mhead->msgsnap_mlen;

		dumpmsg(mhead->msgsnap_mtype, (char *)(mhead + 1), mlen);

		/* advance to next message header */
		/* LINTED alignment */
		mhead = (struct msgsnap_mhead *)
		    ((caddr_t)(mhead + 1) + SZROUND(mlen));
	}
}

/*
 * dumpmsg - dump one message from a message queue.
 */
void
dumpmsg(long type, char *msg, size_t msgsize)
{
	size_t i, j, k;
	int c;

	(void) printf(gettext("  message type %ld, size %lu\n"),
	    type, (ulong_t)msgsize);

	for (i = 0; i < msgsize; i += 16) {
		/* first in hex */
		(void) printf(" %5ld:  ", (ulong_t)i);
		for (j = 0; j < 16; j++) {
			if ((k = i + j) < msgsize)
				(void) printf("%2.2x ", msg[k] & 0xff);
			else
				(void) printf("   ");
		}
		/* then in ascii */
		(void) printf("   ");
		for (j = 0; j < 16; j++) {
			if ((k = i + j) >= msgsize)
				break;
			c = msg[k] & 0xff;
			if (isascii(c) && isprint(c))
				(void) printf("%c", c);
			else
				(void) printf(".");
		}
		(void) printf("\n");
	}
}

/* convert string containing zone name or id to a numeric id */
static zoneid_t
getzone(char *arg)
{
	zoneid_t zoneid;

	if (zone_get_id(arg, &zoneid) != 0) {
		(void) fprintf(stderr,
		    gettext("ipcs: unknown zone: %s\n"), arg);
		exit(1);
	}
	return (zoneid);
}

static void
printzone(zoneid_t id)
{
	char zone_name[ZONENAME_MAX];

	if (getzonenamebyid(id, zone_name, sizeof (zone_name)) < 0)
		(void) printf("%9d", (int)id);
	else
		(void) printf("%9.8s", zone_name);
}
